### ISSUE TITLE
MP Context ThreadContext isn't propagated properly with @RequestScoped

### ISSUE THREAD

**Issue by sdaschner** [quarkusio/quarkus#18958]
### Describe the bug

The `ThreadContext` from MP doesn't reliably capture the context of beans that include a `@RequestScoped` bean.

Not sure if I'm missing something, but the example (that I've tested with other MP impls) described [here](https://blog.sebastian-daschner.com/entries/context-asynchronous-execution-bulkheads-microprofile) doesn't work with Quarkus.

### How to Reproduce?

```java
@ApplicationScoped
@Path("contexts/example")
public class ThreadContextExampleResource {

    @Inject
    ExampleStore exampleStore;

    @Inject
    ThreadContext threadContext;

    @Inject
    ManagedExecutor mes;

    @Inject
    Notifier notifier;

    @PUT
    public void setExample(String example) {
        exampleStore.setExample(example);
        mes.execute(threadContext.contextualRunnable(notifier::notifyAbout));
    }
}

@RequestScoped
class ExampleStore {

    private String example;

    public String getExample() {
        return example;
    }

    public void setExample(String example) {
        this.example = example;
    }
}

@Dependent
class Notifier {

    @Inject
    ExampleStore exampleStore;

    public void notifyAbout() {
        LockSupport.parkNanos(200_000_000L);
        System.out.println("New example: " + exampleStore.getExample());
    }
}
```

Invoking `curl localhost:8080/contexts/example -i -XPUT -H 'Content-Type: text/plain' -d 'foobar'` produces empty logs, if the processing takes longer (our wait for 200ms):

```
New example: null
```

### Quarkus version or git rev

2.0.2.Final

### Additional information

It didn't make a difference whether the `ThreadContext` or `ManagedExecutor` were obtained programmatically or via injection.

**Comment by geoand** [comment]
cc @FroMage

**Comment by manovotn** [comment]
Hi,

can you share the reproducer quarkus app that you used so that we are both looking at identical code?

> produces empty logs, if the processing takes longer (our wait for 200ms):

So, if you don't put in the wait it works as expected?

> `mes.execute(threadContext.contextualRunnable(notifier::notifyAbout));`

You are using both, `ManagedExecutor` and `ThreadContext` - any reason for it? Context will be propagated accordingly to your `ManagedExecutor` settings if you just pass in ordinary executable. `ThreadContext` can be used if you want to pass in settings different from what the executor had. Either way, this shouldn't be what's causing the error.

Might be worth to inspect how is the req. scoped bean created/destroyed - e.g. adding a post construct and  pre-destroy methods with some debug outputs.

**Comment by FroMage** [comment]
I was able to reproduce. What happens, I'm pretty sure, is that the request context gets terminated by the end of the `@PUT` method, so the `ManagedExecutor` starts with a valid request scope, and after its sleep, the request scope is dead, so somehow it gets auto-recreated to an empty value.

This should not work, BTW, because our semantics do say that the request context will be terminated. If you want to throw stuff to an executor, you have to extend the lifetime of the request context by returning a `Uni` or `CompletionStage` from your `@PUT` method, so we know when to terminate it.

At least, this shows the following output:

```java
@ApplicationScoped
@Path("contexts/example")
public class ThreadContextExampleResource {

    @Inject
    ExampleStore exampleStore;

    @Inject
    ThreadContext threadContext;

    @Inject
    ManagedExecutor mes;

    @Inject
    Notifier notifier;

    @PUT
    public void setExample(String example) {
        exampleStore.setExample(example);
        System.err.println("setExample examplestore: "+exampleStore.getId());
        System.err.println("setExample examplestore example: "+example);
        mes.execute(threadContext.contextualRunnable(notifier::notifyAbout));
    }
}


@RequestScoped
class ExampleStore {

    static int ids = 0;
    int id = ids++;
    
    private String example;

    public String getExample() {
        return example;
    }

    public void setExample(String example) {
        this.example = example;
    }

    public String getId() {
        return ""+id;
    }
}

@Dependent
class Notifier {

    @Inject
    ExampleStore exampleStore;

    public void notifyAbout() {
        System.err.println("notifyAbout examplestore: "+exampleStore.getId());
        System.out.println("New example before sleep: " + exampleStore.getExample());
        LockSupport.parkNanos(200_000_000L);
        System.err.println("notifyAbout2 examplestore: "+exampleStore.getId());
        System.out.println("New example after sleep: " + exampleStore.getExample());
    }
}
```

```
setExample examplestore: 1
setExample examplestore example: foobar
notifyAbout examplestore: 1
New example before sleep: null
notifyAbout2 examplestore: 2
New example after sleep: null
```

I'm not sure if my technique is right to figure out which instance of `ExampleStore` we're looking at, with `id`, but the fact that it bumps id in the middle is weird, no?

**Comment by manovotn** [comment]
I find this weird:
```
notifyAbout examplestore: 1
New example before sleep: null
```

How's the example `null` when you had it previously set?

Also, you could add predestroy callback to the req. bean to see when it gets destroyed (and also post construct if you want to see creations). That might give you more hint as to when the bean gets destroyed.

> but the fact that it bumps id in the middle is weird, no?

Yeah, that seems like the context was either re-activated sometime in between or the instances were cleared (i.e. context was deactivated).

**Comment by FroMage** [comment]
```
CONSTRUCT 1
setExample examplestore: 1
setExample examplestore example: foobar
notifyAbout examplestore: 1
DESTROY 1
New example before sleep: foobar
CONSTRUCT 2
notifyAbout2 examplestore: 2
New example after sleep: null
```

So that's it. Occam's razor says it's the `PUT` method destroying the request context at the end of its execution. We do not support propagating request contexts to multiple threads executed in parallel, nor past the request method's lifecycle. So this can't work.

But I think the fact that the request scope gets re-created/activated is weird and perhaps a bug?

**Comment by manovotn** [comment]
> Occam's razor says it's the PUT method destroying the request context at the end of its execution. 

Makes sense because that's just end of single request which is a lifespan of req. context.

> But I think the fact that the request scope gets re-created/activated is weird and perhaps a bug?

I don't know. We are automatically activating req. context for lots of things so this can be a side effect of one of those.

**Comment by sdaschner** [comment]
So, how I understood the MP Context Propagation specs is that this should indeed work to propagate / prolong the request scope when explicitly used with `ThreadContext`. At least, this was the case in other impls, see here: https://blog.sebastian-daschner.com/entries/context-asynchronous-execution-bulkheads-microprofile

For me that was the main reason - besides transactions - to use this functionality, to make use of beans of other scoped that would otherwise be "too small", or not guaranteed to be still active, once the async processing finishes, hence the explicit wait for 200ms...

**Comment by manovotn** [comment]
> So, how I understood the MP Context Propagation specs is that this should indeed work to propagate / prolong the request scope when explicitly used with ThreadContext.

It's not as simple as that. The request context has certain boundaries that spin off the main thread. For example, imagine you are on `Thread1`, where HTTP req. happens and your request context is active. CDI spec defines that the context is active so long as the is the HTTP request. So, if you now propagate this context into `Thread2` and it references the same context with the same beans; what happens when HTTP req. ends?
* Do you destroy the context and all it's beans on both threads? That abruptly kills anything on `Thread2` that is using those beans and still runs
* Do you just clear the context in `Thread1` and "let it continue" in `Thread2`? Well, sounds good but since you are already out of HTTP request, how do you now know when does this context end? It could go on forever.

The first example I gave is pure parallel execution and it cannot be reliably done WRT bean destruction because you then have N threads running tasks for indeterminate amount of time and out of them only one knows when the original context ends but has no information or control over remaining threads. To make it work, you'd have to off load work to other threads, but the original one would have to eventually sync with them and wait.

The latter scenario is (or can be transferred to) a continuous propagation from one thread, that then stops, and continues on in another. This is what Quarkus uses with event loop and other things and here you can easily propagate contexts so long as you also carry the information about when to end - which is the `Uni` or `CompletionStage` that Stephane mentioned.

@FroMage please correct me if I am wrong on some of that^

**Comment by mkouba** [comment]
@manovotn @FroMage So is this really a bug or should we close it as won't fix?

**Comment by manovotn** [comment]
IIRC, the expectations stated in [this comment](https://github.com/quarkusio/quarkus/issues/18958#issuecomment-894002736) are faulty from MP CP (and CDI) perspective. And from the comments I and @FroMage did back then, I'd say we agreed on that.

However, there was some awkward side effect behavior observed that we probably never delved into.
That might not even manifest now as there were multiple changes in CP since this issue was created so I think it makes sense to re-try it and go from there.

**Comment by gsmet** [comment]
@sdaschner could you check if this is still an issue with latest Quarkus? A lot of changes have been made in this area.

**Comment by geoand** [comment]
Any updates on thiu?
